singer_sdk/connectors/aws_boto3.py

```diff
@@ -0,0 +1,272 @@
+        "aws_secret_access_key",
+        th.StringType,
+        secret=True,
+        description="The secret key for your AWS account.",
+    ),
+    th.Property(
+        "aws_session_token",
+        th.StringType,
+        secret=True,
+        description=(
+            "The session key for your AWS account. This is only needed when"
+            " you are using temporary credentials."
+        ),
+    ),
+    th.Property(
+        "aws_profile",
+        th.StringType,
+        description=(
+            "The AWS credentials profile name to use. The profile must be "
+            "configured and accessible."
+        ),
+    ),
+    th.Property(
+        "aws_default_region",
+        th.StringType,
+        description="The default AWS region name (e.g. us-east-1) ",
+    ),
+    th.Property(
+        "aws_endpoint_url",
+        th.StringType,
+        description="The complete URL to use for the constructed client.",
+    ),
+    th.Property(
+        "aws_assume_role_arn",
+        th.StringType,
+        description="The role ARN to assume.",
+    ),
+    th.Property(
+        "use_aws_env_vars",
+        th.BooleanType,
+        default=False,
+        description=("Whether to retrieve aws credentials from environment variables."),
+    ),
+).to_dict()
+
+
+class AWSBoto3Connector:
+    """Base class for AWS boto3-based connectors.
+
+    The connector class serves as a wrapper around boto3 package.
+
+    The functions of the connector are:
+    - initializing a client, resource, or session with a simple interface
+    - accessing AWS credentials via config, env vars, or profile
+    - supports assuming roles
+    - enables configurable endpoint_url for testing
+    """
+
+    def __init__(
+        self: AWSBoto3Connector,
+        config: dict,
+        service_name: str,
+    ) -> None:
+        """Initialize the AWSBotoAuthenticator.
+
+        Args:
+            config (dict): The config for the connector.
+            service_name (str): The name of the AWS service.
+        """
+        self._service_name = service_name
+        self._config = config
+        self._client = None
+        self._resource = None
+        # config for use environment variables
+        if config.get("use_aws_env_vars"):
+            self.aws_access_key_id = os.environ.get("AWS_ACCESS_KEY_ID")
+            self.aws_secret_access_key = os.environ.get("AWS_SECRET_ACCESS_KEY")
+            self.aws_session_token = os.environ.get("AWS_SESSION_TOKEN")
+            self.aws_profile = os.environ.get("AWS_PROFILE")
+            self.aws_default_region = os.environ.get("AWS_DEFAULT_REGION")
+        else:
+            self.aws_access_key_id = config.get("aws_access_key_id")
+            self.aws_secret_access_key = config.get("aws_secret_access_key")
+            self.aws_session_token = config.get("aws_session_token")
+            self.aws_profile = config.get("aws_profile")
+            self.aws_default_region = config.get("aws_default_region")
```

Comment:
Would it make sense for users to want to use a mix of env vars and JSON config? For example, set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` in their environment but switch regions via the config file.

Reply:
@edgarrmondragon potentially, I dont know all the common patterns people use but [my vision](https://github.com/MeltanoLabs/tap-dynamodb/issues/15) for this was to try to make it more explicit where AWS credentials are coming from. I find that having so many variations of where credentials come from (json, profiles, env vars, machine inheritances) and potentially mixing them makes it confusing and sometimes leads to a tap using credentials you weren't intending to use. What do you think about that? Am I exaggerating the challenge a bit?

In tap-dynamodb I was raising an error if credentials weren't provided but some users need to use implicit credentials https://github.com/MeltanoLabs/tap-dynamodb/issues/14. This has me thinking about adding a config option like `use_implicit_credentials=True`, meaning the user needs to explicitly tell the tap that it should use a naked call like `boto3.Session()` with no args/kwargs to defer to the boto hierarchy of searching for credentials on the machine. 

I'm also almost in favor of eliminating the `use_aws_env_vars` option and requiring that all users configure them via their meltano.yml. If they want to access env vars they should use templating in their meltano.yml explicitly.

Any thoughts?